File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/shaders/uniforms2/Mat4Uniform.java
package org.saar.lwjgl.opengl.shaders.uniforms2;

import org.joml.Matrix4fc;
import org.lwjgl.opengl.GL20;
import org.lwjgl.system.MemoryStack;

import java.nio.FloatBuffer;

public abstract class Mat4Uniform extends UniformBase implements Uniform {

    @Override
    public final void doLoad() {
        try (final MemoryStack stack = MemoryStack.stackPush()) {
            final FloatBuffer buffer = stack.mallocFloat(16);
            GL20.glUniformMatrix4fv(getLocation(), isTranspose(),
                    getUniformValue().get(buffer));
        }
    }

    public boolean isTranspose() {
        return false;
    }

    public abstract Matrix4fc getUniformValue();
}

File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/shaders/uniforms/Mat4Uniform.kt
package org.saar.lwjgl.opengl.shaders.uniforms

import org.joml.Matrix4fc
import org.lwjgl.opengl.GL20
import org.lwjgl.system.MemoryStack

abstract class Mat4Uniform : Uniform {

    final override fun load(location: Int) {
        val value: Matrix4fc = this.value
        MemoryStack.stackPush().use { stack ->
            val buffer = value.get(stack.callocFloat(16))
            GL20.glUniformMatrix4fv(location, this.transpose, buffer)
        }
    }

    abstract val value: Matrix4fc
    abstract val transpose: Boolean
}